app/src/test/java/com/kenchen/capstonenewsapp/NewsViewModelTest.kt

```diff
@@ -0,0 +1,145 @@
+        "www.test.com",
+        "CNN",
+        "Show me how to test.",
+    )
+
+    @ExperimentalCoroutinesApi
+    @get:Rule
+    var coroutinesTestRule = CoroutinesTestRule()
+
+    @get:Rule
+    var instantTaskExecutorRule = InstantTaskExecutorRule()
+
+
+    @Before
+    fun setUp() {
+        MockKAnnotations.init(this)
+
+        // FIXME: can we put every in set up before test start?
+//        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+//        viewModel = NewsViewModel(mockRepository)
+    }
+
+    @Test
+    fun `When calling print message actually prints out the message`() {
+
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        assertEquals("This is in test", viewModel.getMessage("This is in test"))
+
+        verify(exactly = 1) { mockRepository.isDataUsage() }
+//        verify(exactly = 1) { viewModel.print("123") }
+    }
+
+    @Test
+    fun `print message`() {
+
+        every { mockRepository.isDataUsage() } answers {
+            flow { }
+//            flow { emit(true) }
+        }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.print("1234")
+
+        verify(exactly = 1) { viewModel.print("1234") }
+//        verify(exactly = 1) { println("1234") }
+    }
+
+    @Test
+    fun `fetch articles`() {
+
+        val mockObserver = mockk<Observer<ArticleState>>()
+
+        val expectedResult = ArticleState.Ready(listOf(dummyArticle))
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        every { mockRepository.getArticles() } returns flow { emit(expectedResult) }
+
+        var onChangedInvoked = 0
+
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.headLineNewsLiveData.observeForever(mockObserver)
+
+        every { mockObserver.onChanged(expectedResult) } answers {
+            onChangedInvoked++
+        }
+
+        viewModel.fetchArticle()
+
+        verify(exactly = 1) { mockRepository.getArticles() }
+
+//        assertEquals(1, onChangedInvoked)
+        verify { mockObserver.onChanged(expectedResult) }
+    }
+
+    @Test
+    fun `search articles`() {
+
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.searchArticles("test")
+
+        coVerify(exactly = 1) { mockRepository.searchArticles(any()) }
+    }
+
+    @Test
+    fun `toggle data usage`() {
+        every { mockRepository.isDataUsage() } returns flow { emit(true) }
+
+        val viewModel = NewsViewModel(mockRepository)
+
+        viewModel.toggleDataUsage()
+
+        coVerify(exactly = 1) { mockRepository.toggleDataUsage() }
+    }
+}
+
+/**
+ * Question:
+ * 1. how do I test the function that returns Unit (nothing)
+ * 2. what is the difference between black-box and white-box testing
+ * 3. can we put all the code in runBlocking ?
```

Comment:
For the most part, yes.  There's a `runTest` block that you can call as well, which arguably is a bit better since it runs all coroutine on a "test" dispatcher and not "main".